Implement per-source-folder module-info for multi-release compilation

laeubi opened the pull request:
See issue 
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/4268

@stephan-herrmann this now works for the *compilation* part but _not for the reconciler_ yet, where I need to check where to inject the info best.

I see two options here:

1. We merge this one and fix the reconciler in another PR (might also be easier to review)
2. We wait until i fixed that as well and delay this a bit more

laeubi committed:
Implement per-source-folder module-info for multi-release compilation

laeubi commented:
> Now there are API errors, I think I'll wait until the next stream opens to fix those...

It seems I totally missed this - sorry for that. @stephan-herrmann should we go on with this to merge it for the next release stream? Was you able to test if this is fixing your JPMS issues you have seen?

laeubi commented:
@stephan-herrmann finally build is green can you review/merge?

laeubi commented:
@stephan-herrmann a friendly reminder for review, it would be good to have this in the M1 to allow extended testing period.

stephan-herrmann commented:
> @stephan-herrmann a friendly reminder for review, it would be good to have this in the M1 to allow extended testing period.

After a large burst of work on type inference I'm back on it now.

stephan-herrmann committed:
polish:
* javadoc of new API
* code layout

stephan-herrmann commented:
For some call paths that are not yet release aware I filed #5149.

The following observation, however, might be good to fix in this PR:

Using the project from the new test case: Go to any of the `Test` classes, then select any of the `Window` or `Element` type references. Try hover, or Go To Declaration (F3 or ctrl+click) and observe the same inconsistency in all versions:
* `Window` is correctly resolved
* `Element` offers a selection of 4 types named "Element"

stephan-herrmann commented:
@laeubi  I tried to polish the javadoc for new API, please double check if anything is amiss now.

For the build failure I'm sure you understand better than me what is going on here (some baseline checks).

laeubi commented:
> Using the project from the new test case: Go to any of the `Test` classes, then select any of the `Window` or `Element` type references. Try hover, or Go To Declaration (F3 or ctrl+click) and observe the same inconsistency in all versions

Is this directly related to classes changed here / module support in general? If not I would prefer to better make smaller PRs than include more than the minimal required to fix one issue at a time.

laeubi commented on org.eclipse.jdt.core.tests.builder/src/org/eclipse/jdt/core/tests/builder/MultiReleaseTests.java:
```diff
@@ -35,6 +35,10 @@ public class MultiReleaseTests extends BuilderTests {
 	private static final String JAVA9_SRC_FOLDER = "src9";
 	private static final String DEFAULT_SRC_FOLDER = "src";
 
+//	static {
+//		TESTS_NAMES = new String[] { "testMultiReleaseModuleInfoPerRelease" };
+//	}
+
 	public MultiReleaseTests(String name) {
```
Is this a left over or intended change?

stephan-herrmann replied:
It's the typical scaffolding for running individual tests, uncomment the block and only your test will run :)

laeubi commented:
> I tried to polish the javadoc for new API, please double check if anything is amiss now.

I added one comment to a change that looks suspicious the rest seems fine so far.

> For the build failure I'm sure you understand better than me what is going on here (some baseline checks).

I think I likely need to re-base the PR if you are fine with that

stephan-herrmann commented:
> > Using the project from the new test case: Go to any of the `Test` classes, then select any of the `Window` or `Element` type references. Try hover, or Go To Declaration (F3 or ctrl+click) and observe the same inconsistency in all versions
> 
> Is this directly related to classes changed here / module support in general? If not I would prefer to better make smaller PRs than include more than the minimal required to fix one issue at a time.

This problem can only be reproduced in an MR project with module-info.java, making the feature introduced here appear incomplete / broken. I haven't analyzed where in the code things go wrong.

stephan-herrmann commented:
> I think I likely need to re-base the PR if you are fine with that

If it's a pure rebase with no further changes hidden inside that's fine.

laeubi commented:
> If it's a pure rebase with no further changes hidden inside that's fine.

I triggered the rebase now, lets see if it helps already...



> This problem can only be reproduced in an MR project with module-info.java, making the feature introduced here appear incomplete / broken.

I can try to take a look but strictly speaking this PR was about *compilation* so possibly still better to have it in an extra PR as it more seem to surface at the UI.

laeubi committed:
Make codeSelect release-aware for multi-release modular projects

 Resolve types and modules from the source folder's own release specific
 module-info.java instead of always using the base module. Fixes
duplicate
 Hover/Open-Declaration results for types only reachable via a higher
 release module-info.

laeubi commented:
`Openable.codeSelect` (the entry point for Hover / Open Declaration) created its `SearchableEnvironment` with `NO_RELEASE`, so it always built the base module (src, which requires nothing) — regardless of which folder's `Test.java` was open. Consequently `org.w3c.dom.Element` is not found in the module graph (`java.xml` not readable), the binding can't be resolved, and `SelectionEngine` falls back to a module-unaware textual `findTypes("Element")` that returns all 4 JDK types named `Element`. `Window` "worked" only because exactly one accessible type is named `Window`.

I tried to add a test and fix the issue for this case now by adding MR awareness to the `Openable.codeSelect` ...

laeubi committed:
Add test for modular reconciling

laeubi committed:
Wire release-awareness into editor name-environment paths

 Make ASTParser binding resolution, code completion, IType.resolveType
and
 type hierarchy resolution honor a release specific module-info.java in
 multi-release modular projects. Consolidate release lookup into shared
 JavaProject.getRelease helpers and add reconcile/AST tests.

laeubi commented:
I now also addressed some other call sites where release awareness was missing. Let me know if you are fine with these or if I should better split up things into individual PRs.

stephan-herrmann commented:
New warning in [ReconcilerModuleMultiReleaseTests.java:109](https://ci.eclipse.org/jdt/job/eclipse.jdt.core-Github/job/PR-4534/16/eclipse/new/source.d3240b46-9829-4607-8a64-e941c5cb42bc/#109): 
```
The method getWorkingCopy(String, String, WorkingCopyOwner, IProblemRequestor) from the type AbstractJavaModelTests is deprecated
```

stephan-herrmann commented on org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/CompilationUnitResolver.java:
```diff
@@ -716,7 +716,11 @@ public static void resolve(
 			int amountOfWork = (compilationUnits.length + bindingKeys.length) * 2; // 1 for beginToCompile, 1 for resolve
 			SubMonitor subMonitor = SubMonitor.convert(monitor, amountOfWork);
-			environment = new CancelableNameEnvironment(((JavaProject) javaProject), owner, subMonitor);
+			// resolve as seen from the source folder the units live in, honoring a release specific module-info.java
+			int release = compilationUnits.length > 0
+					? JavaProject.getRelease(compilationUnits[0])
+					: JavaProject.NO_RELEASE;
+			environment = new CancelableNameEnvironment(((JavaProject) javaProject), owner, subMonitor, false, release);
```
Which test covers this addition? I couldn't find any.

laeubi replied:
Test was added.

stephan-herrmann commented on org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/hierarchy/HierarchyBuilder.java:
```diff
@@ -94,7 +94,8 @@ public HierarchyBuilder(TypeHierarchy hierarchy) throws JavaModelException {
 		}
 		if (project != null) {
-			SearchableEnvironment searchableEnvironment = project.newSearchableNameEnvironment(unitsToLookInside);
+			int release = JavaProject.getRelease(focusType);
```
I couln't find tests covering any of the additions in this package.

laeubi replied:
Test was added.

stephan-herrmann reviewed (commented):
Changes look good to me. Only test coverage seems to be incomplete (or did I miss relevant tests?)

Of course it's good to make more call paths release-aware. So I leave it to you to choose:
* add more test cases here, _or_
* file a follow-up issue to improve test coverage later.

Finally the new compile warning needs to be resolved.

stephan-herrmann commented:
BTW, @laeubi, did you see #5221? While it talks about similar topics, technically it seems to be unrelated, but perhaps worth a look to ensure that the different parts of JDT are consistent with each other (conceptually and at the code level)?

laeubi commented:
> Only test coverage seems to be incomplete (or did I miss relevant tests?)

Its always a bit hard (for me) to guess best ways to cover these especially I'm not always sure what triggers it, so first thing is that I look that it at least does not break anything. Second is that I often try to implement for what we have a non working example, but as you have explicitly asked for other places I thought giving it a try and the pattern is rather repetitive.

> Finally the new compile warning needs to be resolved.

I'll take a look.

laeubi commented:
@stephan-herrmann as the build is now green I think its ready for another review. Let me know if you want me to squash everything into one an/or rebase on current master.

stephan-herrmann commented:
> @stephan-herrmann as the build is now green I think its ready for another review. Let me know if you want me to squash everything into one an/or rebase on current master.

Thanks for adding more tests. Looks good to me now. I performed the squash & rebase now for final testing.
